File: core/src/stats/model/SystemHealth.cpp
#include "stats/model/SystemHealth.hpp"

#include "protocols/ProtocolService.hpp"
#include "protocols/shell/Server.hpp"
#include "runtime/Deps.hpp"
#include "runtime/Manager.hpp"

#include <algorithm>
#include <array>
#include <chrono>
#include <nlohmann/json.hpp>

namespace vh::stats::model {

namespace {

std::uint64_t unixTimestamp() {
    const auto now = std::chrono::system_clock::now();
    return static_cast<std::uint64_t>(std::chrono::system_clock::to_time_t(now));
}

std::size_t countReadyDeps(const DependencyHealth& deps) {
    const std::array<bool, 9> checks{
        deps.storageManager,
        deps.apiKeyManager,
        deps.authManager,
        deps.sessionManager,
        deps.secretsManager,
        deps.syncController,
        deps.fsCache,
        deps.shellUsageManager,
        deps.httpCacheStats
    };

    return static_cast<std::size_t>(std::ranges::count(checks, true));
}

bool depsHealthy(const DependencyHealth& deps, const HealthSummary& summary) {
    return summary.depsReady == summary.depsTotal && deps.fuseSession;
}

bool protocolsHealthy(const ProtocolHealth& protocols) {
    return (!protocols.websocketConfigured || protocols.websocketReady)
        && (!protocols.httpPreviewConfigured || protocols.httpPreviewReady);
}

std::pair<std::size_t, std::size_t> protocolReadySummary(const ProtocolHealth& protocols) {
    std::size_t ready = 0;
    std::size_t total = 0;

    if (protocols.websocketConfigured) {
        ++total;
        if (protocols.websocketReady) ++ready;
    }

    if (protocols.httpPreviewConfigured) {
        ++total;
        if (protocols.httpPreviewReady) ++ready;
    }

    return {ready, total};
}

bool shellHealthy(const ShellHealth& shell) {
    return !shell.adminUidBound.has_value() || *shell.adminUidBound;
}

SystemHealthStatus computeOverallStatus(const SystemHealth& health) {
    if (health.runtime.services.empty())
        return SystemHealthStatus::Critical;

    if (!health.deps.storageManager && !health.deps.authManager && !health.deps.sessionManager)
        return SystemHealthStatus::Critical;

    const bool ok = health.runtime.allRunning
        && protocolsHealthy(health.protocols)
        && depsHealthy(health.deps, health.summary)
        && shellHealthy(health.shell);

    return ok ? SystemHealthStatus::Healthy : SystemHealthStatus::Degraded;
}

}

bool SystemHealth::healthy() const noexcept {
    return overallStatus == SystemHealthStatus::Healthy;
}

std::string SystemHealth::overallStatusString() const {
    return to_string(overallStatus);
}

SystemHealth SystemHealth::snapshot() {
    const auto& manager = runtime::Manager::instance();
    const auto runtimeStatus = manager.status();
    const auto protocolService = manager.getProtocolService();
    const auto shellServer = manager.getShellServer();
    const auto protocolStatus = protocolService
        ? protocolService->protocolStatus()
        : protocols::ProtocolService::RuntimeStatus{};
    const auto depsStatus = runtime::Deps::get().sanityStatus();

    SystemHealth out;
    out.runtime.allRunning = runtimeStatus.allRunning;
    out.runtime.serviceCount = runtimeStatus.services.size();
    out.runtime.services.reserve(runtimeStatus.services.size());

    for (const auto& service : runtimeStatus.services) {
        out.runtime.services.push_back({
            .entryName = service.entryName,
            .serviceName = service.serviceName,
            .running = service.running,
            .interrupted = service.interrupted
        });
    }

    out.protocols = {
        .running = protocolStatus.running,
        .ioContextInitialized = protocolStatus.ioContextInitialized,
        .websocketConfigured = protocolStatus.websocketConfigured,
        .websocketReady = protocolStatus.websocketReady,
        .httpPreviewConfigured = protocolStatus.httpPreviewConfigured,
        .httpPreviewReady = protocolStatus.httpPreviewReady
    };

    out.deps = {
        .storageManager = depsStatus.storageManager,
        .apiKeyManager = depsStatus.apiKeyManager,
        .authManager = depsStatus.authManager,
        .sessionManager = depsStatus.sessionManager,
        .secretsManager = depsStatus.secretsManager,
        .syncController = depsStatus.syncController,
        .fsCache = depsStatus.fsCache,
        .shellUsageManager = depsStatus.shellUsageManager,
        .httpCacheStats = depsStatus.httpCacheStats,
        .fuseSession = depsStatus.fuseSession
    };

    out.shell.adminUidBound = shellServer ? std::optional<bool>(shellServer->adminUIDSet()) : std::nullopt;

    const auto [protocolsReady, protocolsTotal] = protocolReadySummary(out.protocols);
    out.summary = {
        .servicesReady = static_cast<std::size_t>(std::ranges::count_if(
            out.runtime.services,
            [](const RuntimeServiceHealth& service) { return service.running; }
        )),
        .servicesTotal = out.runtime.services.size(),
        .depsReady = countReadyDeps(out.deps),
        .depsTotal = 9,
        .protocolsReady = protocolsReady,
        .protocolsTotal = protocolsTotal,
        .checkedAt = unixTimestamp()
    };

    out.overallStatus = computeOverallStatus(out);
    return out;
}

std::string to_string(const SystemHealthStatus status) {
    switch (status) {
        case SystemHealthStatus::Healthy:
            return "healthy";
        case SystemHealthStatus::Critical:
            return "critical";
        case SystemHealthStatus::Degraded:
        default:
            return "degraded";
    }
}

void to_json(nlohmann::json& j, const RuntimeServiceHealth& health) {
    j = nlohmann::json{
        {"entry_name", health.entryName},
        {"service_name", health.serviceName},
        {"running", health.running},
        {"interrupted", health.interrupted},
    };
}

void to_json(nlohmann::json& j, const RuntimeHealth& health) {
    j = nlohmann::json{
        {"all_running", health.allRunning},
        {"service_count", health.serviceCount},
        {"services", health.services},
    };
}

void to_json(nlohmann::json& j, const ProtocolHealth& health) {
    j = nlohmann::json{
        {"running", health.running},
        {"io_context_initialized", health.ioContextInitialized},
        {"websocket_configured", health.websocketConfigured},
        {"websocket_ready", health.websocketReady},
        {"http_preview_configured", health.httpPreviewConfigured},
        {"http_preview_ready", health.httpPreviewReady},
    };
}

void to_json(nlohmann::json& j, const DependencyHealth& health) {
    j = nlohmann::json{
        {"storage_manager", health.storageManager},
        {"api_key_manager", health.apiKeyManager},
        {"auth_manager", health.authManager},
        {"session_manager", health.sessionManager},
        {"secrets_manager", health.secretsManager},
        {"sync_controller", health.syncController},
        {"fs_cache", health.fsCache},
        {"shell_usage_manager", health.shellUsageManager},
        {"http_cache_stats", health.httpCacheStats},
        {"fuse_session", health.fuseSession},
    };
}

void to_json(nlohmann::json& j, const ShellHealth& health) {
    j = nlohmann::json{
        {"admin_uid_bound", health.adminUidBound ? nlohmann::json(*health.adminUidBound) : nlohmann::json(nullptr)},
    };
}

void to_json(nlohmann::json& j, const HealthSummary& health) {
    j = nlohmann::json{
        {"services_ready", health.servicesReady},
        {"services_total", health.servicesTotal},
        {"deps_ready", health.depsReady},
        {"deps_total", health.depsTotal},
        {"protocols_ready", health.protocolsReady},
        {"protocols_total", health.protocolsTotal},
        {"checked_at", health.checkedAt},
    };
}

void to_json(nlohmann::json& j, const SystemHealth& health) {
    j = nlohmann::json{
        {"overall_status", to_string(health.overallStatus)},
        {"runtime", health.runtime},
        {"protocols", health.protocols},
        {"deps", health.deps},
        {"shell", health.shell},
        {"summary", health.summary},
    };
}

}

File: web/src/models/stats/systemHealth.ts
export type HealthStatus = 'healthy' | 'degraded' | 'critical'

export interface IRuntimeServiceHealth {
  entry_name: string
  service_name: string
  running: boolean
  interrupted: boolean
}

export interface IRuntimeHealth {
  all_running: boolean
  service_count: number
  services: RuntimeServiceHealth[]
}

export interface IProtocolHealth {
  running: boolean
  io_context_initialized: boolean
  websocket_configured: boolean
  websocket_ready: boolean
  http_preview_configured: boolean
  http_preview_ready: boolean
}

export interface IDependencyHealth {
  storage_manager: boolean
  api_key_manager: boolean
  auth_manager: boolean
  session_manager: boolean
  secrets_manager: boolean
  sync_controller: boolean
  fs_cache: boolean
  shell_usage_manager: boolean
  http_cache_stats: boolean
  fuse_session: boolean
}

export interface IShellHealth {
  admin_uid_bound: boolean | null
}

export interface IHealthSummary {
  services_ready: number
  services_total: number
  deps_ready: number
  deps_total: number
  protocols_ready: number
  protocols_total: number
  checked_at: number | string | null
}

export interface ISystemHealth {
  overall_status: HealthStatus
  runtime: RuntimeHealth
  protocols: ProtocolHealth
  deps: DependencyHealth
  shell: ShellHealth
  summary: HealthSummary
}

function asObject(v: unknown): Record<string, unknown> {
  return v && typeof v === 'object' && !Array.isArray(v) ? (v as Record<string, unknown>) : {}
}

function asString(v: unknown, fallback = ''): string {
  return typeof v === 'string' ? v : fallback
}

function asNumber(v: unknown, fallback = 0): number {
  return typeof v === 'number' && Number.isFinite(v) ? v : fallback
}

function asBoolean(v: unknown, fallback = false): boolean {
  return typeof v === 'boolean' ? v : fallback
}

function asNullableBoolean(v: unknown): boolean | null {
  return typeof v === 'boolean' ? v : null
}

function asHealthStatus(v: unknown): HealthStatus {
  return v === 'healthy' || v === 'critical' || v === 'degraded' ? v : 'degraded'
}

function asCheckedAt(v: unknown): number | string | null {
  if (typeof v === 'number' && Number.isFinite(v)) return v
  if (typeof v === 'string' && v.length > 0) return v
  return null
}

function countReady(values: boolean[]): number {
  return values.filter(Boolean).length
}

export class RuntimeServiceHealth implements IRuntimeServiceHealth {
  entry_name = 'unknown'
  service_name = 'unknown'
  running = false
  interrupted = false

  constructor(data?: Partial<IRuntimeServiceHealth>) {
    if (!data) return
    this.entry_name = data.entry_name ?? this.entry_name
    this.service_name = data.service_name ?? this.service_name
    this.running = data.running ?? this.running
    this.interrupted = data.interrupted ?? this.interrupted
  }

  static from(raw: unknown): RuntimeServiceHealth {
    const data = asObject(raw)
    return new RuntimeServiceHealth({
      entry_name: asString(data.entry_name, 'unknown'),
      service_name: asString(data.service_name, 'unknown'),
      running: asBoolean(data.running),
      interrupted: asBoolean(data.interrupted),
    })
  }
}

export class RuntimeHealth implements IRuntimeHealth {
  all_running = false
  service_count = 0
  services: RuntimeServiceHealth[] = []

  constructor(data?: Partial<IRuntimeHealth>) {
    if (!data) return
    this.all_running = data.all_running ?? this.all_running
    this.services = data.services ?? this.services
    this.service_count = data.service_count ?? this.services.length
  }

  static from(raw: unknown): RuntimeHealth {
    const data = asObject(raw)
    const services = Array.isArray(data.services) ? data.services.map(RuntimeServiceHealth.from) : []
    return new RuntimeHealth({
      all_running: asBoolean(data.all_running),
      service_count: asNumber(data.service_count, services.length),
      services,
    })
  }
}

export class ProtocolHealth implements IProtocolHealth {
  running = false
  io_context_initialized = false
  websocket_configured = false
  websocket_ready = false
  http_preview_configured = false
  http_preview_ready = false

  constructor(data?: Partial<IProtocolHealth>) {
    if (!data) return
    this.running = data.running ?? this.running
    this.io_context_initialized = data.io_context_initialized ?? this.io_context_initialized
    this.websocket_configured = data.websocket_configured ?? this.websocket_configured
    this.websocket_ready = data.websocket_ready ?? this.websocket_ready
    this.http_preview_configured = data.http_preview_configured ?? this.http_preview_configured
    this.http_preview_ready = data.http_preview_ready ?? this.http_preview_ready
  }

  static from(raw: unknown): ProtocolHealth {
    const data = asObject(raw)
    return new ProtocolHealth({
      running: asBoolean(data.running),
      io_context_initialized: asBoolean(data.io_context_initialized),
      websocket_configured: asBoolean(data.websocket_configured),
      websocket_ready: asBoolean(data.websocket_ready),
      http_preview_configured: asBoolean(data.http_preview_configured),
      http_preview_ready: asBoolean(data.http_preview_ready),
    })
  }

  configuredTotal(): number {
    return Number(this.websocket_configured) + Number(this.http_preview_configured)
  }

  readyTotal(): number {
    return (
      Number(this.websocket_configured && this.websocket_ready)
      + Number(this.http_preview_configured && this.http_preview_ready)
    )
  }
}

export class DependencyHealth implements IDependencyHealth {
  storage_manager = false
  api_key_manager = false
  auth_manager = false
  session_manager = false
  secrets_manager = false
  sync_controller = false
  fs_cache = false
  shell_usage_manager = false
  http_cache_stats = false
  fuse_session = false

  constructor(data?: Partial<IDependencyHealth>) {
    if (!data) return
    this.storage_manager = data.storage_manager ?? this.storage_manager
    this.api_key_manager = data.api_key_manager ?? this.api_key_manager
    this.auth_manager = data.auth_manager ?? this.auth_manager
    this.session_manager = data.session_manager ?? this.session_manager
    this.secrets_manager = data.secrets_manager ?? this.secrets_manager
    this.sync_controller = data.sync_controller ?? this.sync_controller
    this.fs_cache = data.fs_cache ?? this.fs_cache
    this.shell_usage_manager = data.shell_usage_manager ?? this.shell_usage_manager
    this.http_cache_stats = data.http_cache_stats ?? this.http_cache_stats
    this.fuse_session = data.fuse_session ?? this.fuse_session
  }

  static from(raw: unknown): DependencyHealth {
    const data = asObject(raw)
    return new DependencyHealth({
      storage_manager: asBoolean(data.storage_manager),
      api_key_manager: asBoolean(data.api_key_manager),
      auth_manager: asBoolean(data.auth_manager),
      session_manager: asBoolean(data.session_manager),
      secrets_manager: asBoolean(data.secrets_manager),
      sync_controller: asBoolean(data.sync_controller),
      fs_cache: asBoolean(data.fs_cache),
      shell_usage_manager: asBoolean(data.shell_usage_manager),
      http_cache_stats: asBoolean(data.http_cache_stats),
      fuse_session: asBoolean(data.fuse_session),
    })
  }

  coreValues(): boolean[] {
    return [
      this.storage_manager,
      this.api_key_manager,
      this.auth_manager,
      this.session_manager,
      this.secrets_manager,
      this.sync_controller,
      this.fs_cache,
      this.shell_usage_manager,
      this.http_cache_stats,
    ]
  }
}

export class ShellHealth implements IShellHealth {
  admin_uid_bound: boolean | null = null

  constructor(data?: Partial<IShellHealth>) {
    if (!data) return
    this.admin_uid_bound = data.admin_uid_bound ?? this.admin_uid_bound
  }

  static from(raw: unknown): ShellHealth {
    const data = asObject(raw)
    return new ShellHealth({ admin_uid_bound: asNullableBoolean(data.admin_uid_bound) })
  }
}

export class HealthSummary implements IHealthSummary {
  services_ready = 0
  services_total = 0
  deps_ready = 0
  deps_total = 9
  protocols_ready = 0
  protocols_total = 0
  checked_at: number | string | null = null

  constructor(data?: Partial<IHealthSummary>) {
    if (!data) return
    this.services_ready = data.services_ready ?? this.services_ready
    this.services_total = data.services_total ?? this.services_total
    this.deps_ready = data.deps_ready ?? this.deps_ready
    this.deps_total = data.deps_total ?? this.deps_total
    this.protocols_ready = data.protocols_ready ?? this.protocols_ready
    this.protocols_total = data.protocols_total ?? this.protocols_total
    this.checked_at = data.checked_at ?? this.checked_at
  }

  static from(raw: unknown, runtime: RuntimeHealth, deps: DependencyHealth, protocols: ProtocolHealth): HealthSummary {
    const data = asObject(raw)
    return new HealthSummary({
      services_ready: asNumber(data.services_ready, runtime.services.filter(service => service.running).length),
      services_total: asNumber(data.services_total, runtime.service_count),
      deps_ready: asNumber(data.deps_ready, countReady(deps.coreValues())),
      deps_total: asNumber(data.deps_total, deps.coreValues().length),
      protocols_ready: asNumber(data.protocols_ready, protocols.readyTotal()),
      protocols_total: asNumber(data.protocols_total, protocols.configuredTotal()),
      checked_at: asCheckedAt(data.checked_at),
    })
  }
}

export class SystemHealth implements ISystemHealth {
  overall_status: HealthStatus = 'degraded'
  runtime = new RuntimeHealth()
  protocols = new ProtocolHealth()
  deps = new DependencyHealth()
  shell = new ShellHealth()
  summary = new HealthSummary()

  constructor(data?: Partial<ISystemHealth>) {
    if (!data) return
    this.overall_status = data.overall_status ?? this.overall_status
    this.runtime = data.runtime ?? this.runtime
    this.protocols = data.protocols ?? this.protocols
    this.deps = data.deps ?? this.deps
    this.shell = data.shell ?? this.shell
    this.summary = data.summary ?? this.summary
  }

  static from(raw: unknown): SystemHealth {
    const data = asObject(raw)
    const runtime = RuntimeHealth.from(data.runtime)
    const protocols = ProtocolHealth.from(data.protocols)
    const deps = DependencyHealth.from(data.deps)
    const shell = ShellHealth.from(data.shell)
    const summary = HealthSummary.from(data.summary, runtime, deps, protocols)

    return new SystemHealth({
      overall_status: asHealthStatus(data.overall_status),
      runtime,
      protocols,
      deps,
      shell,
      summary,
    })
  }
}
